worker/worker.go

```diff
@@ -902,26 +902,32 @@ func (w *worker) slabMigrateHandler(jc jape.Context) {
 		return
 	}
 
-	err = migrateSlab(ctx, w.downloadManager, w.uploadManager, &slab, dlContracts, ulContracts, up.CurrentHeight, w.logger)
+	// migrate the slab
+	used, err := migrateSlab(ctx, w.downloadManager, w.uploadManager, &slab, dlContracts, ulContracts, up.CurrentHeight, w.logger)
 	if jc.Check("couldn't migrate slabs", err) != nil {
 		return
 	}
 
-	usedContracts := make(map[types.PublicKey]types.FileContractID)
+	// build host to contract map
+	h2c := make(map[types.PublicKey]types.FileContractID)
+	for _, contract := range ulContracts {
+		h2c[contract.HostKey] = contract.ID
+	}
+
 	for _, ss := range slab.Shards {
-		if _, exists := usedContracts[ss.Host]; exists {
+		if _, exists := used[ss.Host]; exists {
 			continue
 		}
 
-		for _, c := range ulContracts {
-			if c.HostKey == ss.Host {
-				usedContracts[ss.Host] = c.ID
-				break
-			}
+		fcid, exists := h2c[ss.Host]
+		if !exists {
+			jc.Error(fmt.Errorf("couldn't find contract for host %v", ss.Host), http.StatusInternalServerError)
+			continue
 		}
+		used[ss.Host] = fcid
```

Comment:
Wait why is that? Can't you directly pass `used` to `UpdateSlab` instead of looping over `slab.Shards` again?

Reply:
no you can't because it'll only contain the used contracts for the migrated shards, I updated the code though so `migrateSlab` takes care of that for you

Reply:
Ok in that case can we rename `Migrate` to `UploadShards`? It was quite confusing to me that in one place we call `Migrate` and are able to return `used` without modification and in another we call `migrateSlabs` and we still have to merge the maps.
Only to realise that `Migrate` isn't actually migrating but only uploading.